src/main/java/seedu/address/model/student/UniqueStudentList.java

```diff
@@ -200,38 +258,224 @@ create a class from the start time and check to see if it exceeds the startTime
+                    ? true
+                    : endTimeFromSecondClass.compareTo(aThirdClass.startTime) <= 0;
+            boolean isNotClashWithThirdClassFromCurrTime = aThirdClass == null
+                    ? true
+                    : endTimeFromCurrTime.compareTo(aThirdClass.startTime) <= 0;
+
+            // boolean which checks whether the endTime is at the end of the day, it if it, then it is at the end of
+            // the day and no timeRange can be larger than it
+            boolean isSecondClassEndOfDay = aSecondClass.endTime.equals(LocalTime.of(0, 0));
+            boolean isFirstClassEndOfDay = aFirstClass.endTime.equals(LocalTime.of(0, 0));
 
-            assert aFirstClass.date != null;
             if (aFirstClass.date.equals(aSecondClass.date)) {
-                /*
-                 * That means they are on the same day
-                 * 1st case: When they are side by side. Since the case where it is before the class has been handled,
-                 *           we try finding a slot after the end of the secondClass.
-                 * 2nd case: When there is a gap, but it is not big enough. If there is a gap, then it is actually
-                 *           the same situation as the first case so, it becomes <= tr.duration rather than just == 0
-                 *           for the initial first case.
-                 * 3rd case: When there is a gap just nice or too big
-                 */
-                assert aFirstClass.endTime != null;
-                assert aSecondClass.startTime != null;
-                if (aFirstClass.endTime.until(aSecondClass.startTime, ChronoUnit.MINUTES) > tr.duration) {
-                    // you are now handling the 3rd case, the 1st case does not matter since if it is outside of the
-                    // duration, it is the next iteration's problem
-                    newClass = new Class(aFirstClass.date, aFirstClass.endTime,
-                            aFirstClass.endTime.plusMinutes(tr.duration));
-                    break;
+                if (isFirstClassEndOfDay || isSecondClassEndOfDay) {
+                    continue;
+                }
+
+                if (aFirstClass.date.equals(currDate)) {
+                    if (currTime.compareTo(tr.startTimeRange) <= 0 && i == 0) {
+                        // before the tr.startTimeRange
+                        if (isGapBetweenClassesLargerThanDuration(aFirstClass, aSecondClass, tr.duration)) {
+                            // the gap is larger than the duration, so no need to check whether if
+                            // endTimeFromFirstClass.compareTo(startTimeFromSecondClass) <= 0
+                            if (tr.startTimeRange.compareTo(aSecondClass.endTime) >= 0) {
+                                newClass = new Class(currDate, startTimeFromTr, endTimeFromTr);
+                            } else if (aSecondClass.startTime.compareTo(tr.startTimeRange) <= 0) {
```

Comment:
variable naming can be improved to be `startTimeFromSecondClass`. Same applies for the rest.

Reply:
Where is this referring to?

Reply:
> variable naming can be improved to be `startTimeFromSecondClass`. Same applies for the rest.

I have realised that the above-declared variables are specially targeted towards the class that is to be created instead of the literal meaning as I interpreted earlier.

Reply:
> Where is this referring to?

I think my initial issue has been resolved after carefully examining your code. I think we can work on the style later after sorting through the logic.